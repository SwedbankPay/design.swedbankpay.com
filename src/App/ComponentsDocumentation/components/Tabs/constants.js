import React from "react";
import Tag from "@components/Tag";
import Tabs from "@components/Tabs";

const tabItems = ["Selected", "Unselected", "Unselected", "Unselected", "Unselected", "Unselected"];

const TabsExample = ({ scroll, id }) => (
    <div className="w-75">
        {scroll === true
            ? <><Tabs id={id} items={tabItems} scroll="true"/></>
            : <Tabs id={id} items={tabItems.slice(0, 3)} />
        }
        <p>Content of the selected tabs</p>
    </div>
);

export const tabsShowCasePanel = {
    id: "no-tabs",
    elements: [
        {
            component: <TabsExample id="tabs-showcase-example"/>,
            options: {
                checkbox: [
                    {
                        title: "Behavior",
                        inputs: [
                            {
                                id: "scroll",
                                name: "Scrollable",
                                value: {
                                    scroll: true
                                }
                            }

                        ]
                    }
                ]
            },
            title: "Tabs",
            description: <>Tabs appear in a single row, above their content. Only one tab can be selected at a time and the selected tab decides which content that is shown in the container beneath. The width of each tab is dependent on its label and the width of the row of tabs is dependent on the container. If the container canâ€™t display all the tabs at once the tabs can be set to scrollable by adding the class <Tag code text=".tabs-scroll" type="secondary"/>, but it is not optimized for desktop usage at the moment.</>
        }
    ]
};
