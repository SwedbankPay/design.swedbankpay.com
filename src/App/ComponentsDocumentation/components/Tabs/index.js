import React, { Component } from "react";

import { ComponentPreview, DocContainer, JavascriptDocs, JavascriptDocElement } from "@docutils";
import TabsComponent from "@components/Tabs";
import CodeTags from "@components/CodeTags";

import { tabs } from "@src/scripts/main";

const tabItems = ["Selected", "Unselected", "Unselected", "Unselected", "Unselected", "Unselected"];

const Overview = () => (
    <section>
        <h2 id="overview">Overview</h2>
        <div className="showcase-panel p-5">
            <div className="w-75 m-auto">
                <TabsComponent id="tabs-example" items={[...tabItems.slice(0, 3)]}/>
            </div>
        </div>
        <div className="component-description border-left border-right p-4 pb-2">
            <span className="h3 d-inline-block mb-2">Tabs</span>
            <p className="mb-0">
                The tabs component has three different states, the selected state for the category currently displayed. Unselected state for all the other categories not currently selected and lastly there the hover state when the user hover over an unselected state.
            </p>
        </div>
        <ComponentPreview language="html" codeFigure>
            <TabsComponent id="tabs-example" items={[...tabItems.slice(0, 3)]}/>
        </ComponentPreview>

        <section>
            <h3>When to consider something else</h3>
            <ul className="list list-bullet">
                <li>It is important to consider whether or not dividing the content into sub-categories is helpful for the user. Using Tabs isn’t always the best solution and can create more fragmented experience and leading to issues where the user might miss out on content. Similarly, this is also not idea if the user needs to compare the content. In those cases it is better to present the content on a page with proper headlines and structure.</li>
            </ul>
        </section>

        <section>
            <h3>Scrollable tabs</h3>
            <div className="slab slab-plain">
                <div className="w-75 m-auto">
                    <TabsComponent id="tabs-scroll-example" ulId="tabs-scroll-example-ul" items={tabItems} scroll/>
                </div>
            </div>
            <p>The tab component also includes scrollable behaviour with the addition of a clickable chevron arrow when the size of the viewport no longer can display all the tab categories at once. The user can navigate inside the tabs menu.</p>
        </section>

        <section id="how-to-use-tabs">
            <h3>How to use tabs</h3>
            <div className="slab slab-plain slab-border-success">
                <span className="h3">Do</span>
                <div className="px-5 py-3">
                    <TabsComponent id="tabs-scroll-example" ulId="tabs-scroll-example-ul" items={[...tabItems.slice(0, 3)]}/>
                    <div className="slab slab-default p-5"></div>
                </div>
            </div>
            <div className="slab slab-plain slab-border-error">
                <span className="h3">Don&apos;t</span>
                <div className="px-5 py-3">
                    <TabsComponent id="tabs-scroll-example" ulId="tabs-scroll-example-ul" items={[...tabItems.slice(0, 3)]} scroll/>
                    <div className="slab slab-default p-5"></div>
                </div>
            </div>
            <p>The tab component should always fill the full container of the content it governs. As seen on the examples above, the length of the underline for tab isn’t determined by the amount of tabs/length of the text. But should instead indicate the content that it anchors to and contains.</p>
        </section>
    </section>
);

const ContentGuidelines = () => (
    <section>
        <h2 id="content-guidelines">Content guidelines</h2>
        <div className="slab slab-plain px-5">
            <div className="w-75 m-auto">
                <TabsComponent id="tabs-scroll-example" ulId="tabs-scroll-example-ul" items={["Selected Label", "Unselected Label", "Unselected Label"]}/>
            </div>
        </div>
        <ul className="list list-bullet">
            <li>The <b>Label</b> for tabs should state the name of the category and division. This should be short and to the point making is as easy as possible for the user to find their desired category.</li>
            <li>The <b>order of labels</b> should always be arranged according the content, sorted after relevance. However, in cases where there is not a preferred order, the tabs should be arrange the tabs in alphabetical order instead.</li>
        </ul>
    </section>
);

const SetScrollStateJavaScript = ({ componentName }) => (
    <>
        <JavascriptDocElement
            code={<CodeTags type="secondary" code={`dg.${componentName}.setScrollState(<${componentName.toLowerCase()}-id>, scrollState)`} />}
            description={
                <>
                    Moves the scroll position of the {componentName} to the user specified position. The value to be passed to <CodeTags type="secondary" code="scrollState" /> is
                    the object <CodeTags type="secondary" code="{ scrollStart, scrollTotalAmount }" />. <CodeTags type="secondary" code="scrollStart" /> is the current scroll
                    position, <CodeTags type="secondary" code="scrollTotalAmount" /> is the amount to be scrolled from the current scroll position (negative values for left scroll,
                    positive values for right scroll). Note: <CodeTags type="primary" code={"<ul>"} /> is the scrollable element.
                </>
            }
        />
    </>
);

const DeveloperDocumentation = () => (
    <>
        <h2 id="developer-documentation">Developer documentation</h2>
        <span className="h3">Javascript methods</span>
        <JavascriptDocs componentName="tabs" others={[SetScrollStateJavaScript]} />
    </>
);

class Tabs extends Component {
    componentDidMount () {
        tabs.init();
    }

    render () {
        return (
            <DocContainer docToc>
                <p className="lead">
                Tabs component separates content into groups within a single content, it lets the user navigate and quickly switch between section within the same content.
                </p>
                <Overview />
                <ContentGuidelines />
                <DeveloperDocumentation />
            </DocContainer>
        );
    }
}

export default Tabs;

/* For testing */
export { Overview, ContentGuidelines, DeveloperDocumentation };
