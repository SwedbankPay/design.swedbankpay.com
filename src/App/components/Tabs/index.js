import React, { Component } from "react";
import PropTypes from "prop-types";

class Tabs extends Component {
    constructor (props) {
        super(props);

        this.state = { active: 0 };
    }

    setActive (e, i) {
        e.preventDefault();
        this.setState({ active: i });
    }

    render () {
        const { id, scroll, items, ulId } = this.props;

        return (
            <div id={id} className={`tabs${scroll ? " tabs-scroll" : ""}`}>{"\n"}
                <ul id={ulId}>
                    {items.map((name, i) => (
                        <li key={`tab-item-${name}-${i}`} className={this.state.active === i ? "active" : null}>{"\n"}
                            <a href="#" onClick={e => this.setActive(e, i)}>{name}</a>{"\n"}
                        </li>
                    ))}
                </ul>
            </div>
        );
    }
}

Tabs.propTypes = {
    id: PropTypes.string,
    items: PropTypes.array.isRequired,
    scroll: PropTypes.bool,
    ulId: PropTypes.string
};

export default Tabs;
