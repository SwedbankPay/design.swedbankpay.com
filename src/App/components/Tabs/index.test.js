import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event/";

import Tabs from "./index";

const items = ["test", "test 2"];
const div = document.createElement("div");

document.body.appendChild(div);

describe("Component: Tabs -", () => {
	it("is defined", () => {
		expect(Tabs).toBeDefined();
	});

	it("renders", () => {
		const componentForSnap = renderer.create(<Tabs items={items} />);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});

	it("renders with class name tabs when prop scroll is not specified", () => {
		const { container } = render(<Tabs items={items} />);

		expect(container.querySelector(".tabs")).not.toHaveClass("tabs-scroll");

		const componentForSnap = renderer.create(<Tabs items={items} />);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});

	it("renders with class name .tabs.tabs-scroll when prop scroll is specified", () => {
		const { container } = render(<Tabs items={items} scroll />);

		expect(container.querySelector(".tabs")).toHaveClass("tabs-scroll");

		const componentForSnap = renderer.create(<Tabs items={items} scroll />);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});

	it("renders with id when id is provided", () => {
		const { container } = render(<Tabs id="test" items={items} />);

		expect(container.querySelector(".tabs")).toHaveAttribute("id", "test");
	});

	it("renders with ulId when ulId is provided", () => {
		render(<Tabs ulId="test" items={items} />);

		expect(screen.getByRole("list")).toHaveAttribute("id", "test");
	});

	it("sets a tab to active", () => {
		render(<Tabs items={items} />);

		expect(
			screen
				.getAllByRole("listitem")
				.some((elmt) => elmt.classList.contains("active"))
		).toBeTruthy();
	});

	// FIXME: click is not working yet. Anyway, it is only testing the React component and its functionality, here we are NOT testing the actual `tabs` script
	it("doesn't remove active when clicking an active tab", () => {
		render(<Tabs items={items} />);

		expect(
			screen
				.getAllByRole("listitem")
				.filter((elmt) => elmt.classList.contains("active"))
		).toHaveLength(1);

		const activeTab = screen
			.getAllByRole("listitem")
			.filter((elmt) => elmt.classList.contains("active"))[0];

		expect(activeTab).toBeTruthy();

		userEvent.click(activeTab);

		// expect(activeTab.classList).toContain("active");
	});

	// FIXME: click is not working yet. Anyway, it is only testing the React component and its functionality, here we are NOT testing the actual `tabs` script.
	it("changes active tab when unactive tab is clicked", () => {
		render(<Tabs items={items} />);

		const inactiveTabs = screen
			.getAllByRole("listitem")
			.filter((elmt) => !elmt.classList.contains("active"));
		const inactiveTab = inactiveTabs[0];

		expect(inactiveTabs.length).toBeGreaterThan(0);

		const tabAnchor = inactiveTab.querySelector("a");

		expect(tabAnchor).toBeTruthy();

		userEvent.click(tabAnchor);

		// expect(inactiveTab).toHaveClass("active");
	});
});
