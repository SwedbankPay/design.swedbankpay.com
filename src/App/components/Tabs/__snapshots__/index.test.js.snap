// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Tabs - renders 1`] = `
<div
  className="tabs"
>
  

  <ul>
    <li
      className="active"
      key="tab-item-test-0"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test
      </a>
      

    </li>
    <li
      className={null}
      key="tab-item-test 2-1"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test 2
      </a>
      

    </li>
  </ul>
</div>
`;

exports[`Component: Tabs - renders with class name .tabs.tabs-scroll when prop scroll is specified 1`] = `
<div
  className="tabs tabs-scroll"
>
  

  <ul>
    <li
      className="active"
      key="tab-item-test-0"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test
      </a>
      

    </li>
    <li
      className={null}
      key="tab-item-test 2-1"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test 2
      </a>
      

    </li>
  </ul>
</div>
`;

exports[`Component: Tabs - renders with class name tabs when prop scroll is not specified 1`] = `
<div
  className="tabs"
>
  

  <ul>
    <li
      className="active"
      key="tab-item-test-0"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test
      </a>
      

    </li>
    <li
      className={null}
      key="tab-item-test 2-1"
    >
      

      <a
        href="#"
        onClick={[Function]}
      >
        test 2
      </a>
      

    </li>
  </ul>
</div>
`;
